internal/ingress/controller/controller.go

```diff
@@ -271,18 +296,22 @@ func (n *KongController) Stop() error {
 		return fmt.Errorf("shutdown already in progress")
 	}
 
-	glog.Infof("shutting down controller queues")
-	close(n.stopCh)
-	go n.syncQueue.Shutdown()
 	if n.syncStatus != nil {
+		// Finish writing to any Ingress objects before giving up leadership.
 		n.syncStatus.Shutdown(n.elector.IsLeader())
 	}
+	glog.Infof("shutting down controller queues")
+	// Closing the stop channel will cause us to give up leadership.
+	close(n.stopCh)
+	n.syncQueue.Shutdown()
```

Comment:
These two lines need to flip. If we close the `n.stop` channel before stopping the queue, the worker goroutine terminates, and the `Shutdown` method never completes, waiting on a channel to be closed. If instead we call `Shutdown` on the queue first, the worker goroutine sees that the queue is shutting down and stops consuming items.